contracts/Leverage.sol

```diff
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: Unlicense
+
+pragma solidity ^0.8.0;
+
+import "hardhat/console.sol";
+import "./interfaces/Compound.sol";
+import "./CompoundMiddleContract.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract Leverage {
+    using SafeERC20 for IERC20;
+
+    /**
+     * @dev leverages the supplied ether by supplying the amount, borrowing max. possible amount to the supply and supplying it again
+     * @param _cEtherAddress address of cEther contract
+     * @param _comptrollerAddress address of Compound comptroller contract
+     * @param _middleContractAddress address of middle contract (to be able to call borrow, deposit functions)
+     */
+    function leverageEther(
+        address payable _cEtherAddress, 
+        address _comptrollerAddress, 
+        address payable _middleContractAddress
+    ) external payable {
+        CompoundMiddleContract middle = CompoundMiddleContract(_middleContractAddress);
+        uint256 supplyEth = msg.value;
+        console.log(supplyEth);
+        // supply ether
+        bool depositSuccess = middle.depositEth{value: msg.value}(_cEtherAddress);
+
+        require(depositSuccess == true, "LEVERAGE: DEPOSIT FAILED");
+        
+        // calculate maximum borrowable amount for the supplied ether
+        Comptroller comptroller = Comptroller(_comptrollerAddress);
+        (, uint collateralFactorMantissa, ) = comptroller.markets(_cEtherAddress);
+        uint256 borrowableEthAmount = (collateralFactorMantissa - (1*(10**17))) * supplyEth / 10**18;
+        console.log("collateralFactorMantissa: ", (collateralFactorMantissa - (1*(10**17))));
+        // borrow the amount
+        middle.borrowEth(_cEtherAddress, _comptrollerAddress, _cEtherAddress, borrowableEthAmount, false, payable(address(this)));
+
```

Comment:
Why are we leveraging two times?

Reply:
If the debt-collateral ratio reaches the maximum threshold, then the whole tx will fail. It would be better to either do single leverage for simplicity or check whether the debt-collateral ratio is below the maximum threshold?

Reply:
What is meant by leveraging two times? The complete process of supplying ETH, then borrowing against that supplied ETH , and finally supplying the borrowed ETH again is one leverage right?

Reply:
Supply -> borrow -> supply

Reply:
In the main contract, you supplied some ETH; then in `borrowEth,` you borrowed some ETH.
Then in the `leverage` contract, you again supplied the ETH in line 29. This flow is correct.

But then again, in line 39, you again borrowed why?

Reply:
So this, flow is kind of independent of the previous positions the contract might have had in Compound

Reply:
Then it would be better to add this functionality separately, not in the `deposit` function!

Reply:
> Then it would be better to add this functionality separately, not in the `deposit` function!

Did not get this 😅

Reply:
> Also, Let's return the Debt-collateral ratio at the end!

Okay